rest-client/src/main/java/org/sagebionetworks/bridge/rest/ApiClientProvider.java

```diff
@@ -26,23 +33,122 @@
     private static final Interceptor WARNING_INTERCEPTOR = new WarningHeaderInterceptor();
     private static final Interceptor ERROR_INTERCEPTOR = new ErrorResponseInterceptor();
     private static final Interceptor LOGGING_INTERCEPTOR = new LoggingInterceptor();
-    
+
+    private static final int NUM_USERS_TO_CACHE = 5;
+    private static final int NUM_SERVICES_PER_USER_TO_CACHE = 15;
+
     private final String baseUrl;
     private final String userAgent;
     private final String acceptLanguage;
-    private final Map<SignIn,UserSessionInfoProvider> sessionProviders;
+    @Deprecated
+    private final Map<SignIn,UserSessionInfoProvider> signInSessionProviders;
+    private final ConcurrentMap<String, UserSessionInfoProvider> emailSessionProviders;
+    private final RetrofitFactory retrofitFactory;
+
+    private final LoadingCache<UserSessionInfoProvider, Retrofit> authenticatedRetrofits;
```

Comment:
Reintroduced caching with Guava's LoadingCache instead of the Map<K,WeakReference<V>> we had before. 

Alternately, I could see ApiClientProvider having a builder pattern to produce an instance bound to an account (builder could take in phone, email-only, email+password).  This would work well on Android and not require much change, haven't looked at the other rest-client consumers. A nice-to-have feature would be to add OkHttp3 interceptors.

In OkHttp3, clients each  have a private connection pool instead of sharing a global pool.
http://square.github.io/okhttp/3.x/okhttp/okhttp3/OkHttpClient.html

> There is no longer a global singleton connection pool. In OkHttp 2.x, all OkHttpClient instances shared a common connection pool by default. In OkHttp 3.x, each new OkHttpClient gets its own private connection pool. Applications should avoid creating many connection pools as doing so prevents connection reuse. Each connection pool holds its own set of connections alive so applications that have many pools also risk exhausting memory!

> OkHttp performs best when you create a single OkHttpClient instance and reuse it for all of your HTTP calls. This is because each client holds its own connection pool and thread pools. Reusing connections and threads reduces latency and saves memory. Conversely, creating a client for each request wastes resources on idle pools.

I've read similar things about Retrofit, but not as strong of an explanation. I do know Reflection on Android is known to be really slow though...

Reply:
It seems like we're doing two things here: 1) caching retrofit clients and 2) fixing the indexing of UserSessionInfoProviders by a sign in object. I should also point out that I created ClientManager specifically because ApiClientProvider was caching different sign ins (ClientManager maps 1:1 with sign in credentials, that create an ApiClientProvider that's only ever going to have one UserSessionInfoProvider in its map). I did it this way to not break existing uses of ApiClientProvider.

I would probably start by fixing the fact that we have a number of credentials under which sessions might be created: email, phone, email+password, and phone+password (all scoped to a study).